src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs

```diff
@@ -17,23 +17,29 @@ internal abstract partial class AbstractPullDiagnosticHandler<TDiagnosticsParams
 
+    internal record struct CheapDiagnosticState(int GlobalStateVersion, SourceGeneratorExecutionVersion ExecutionVersion, VersionStamp DependentVersion);
+
+    internal record struct ExpensiveDiagnosticState(int GlobalStateVersion, SourceGeneratorExecutionVersion ExecutionVersion, Checksum DependentChecksum);
+
     /// <summary>
     /// Cache where we store the data produced by prior requests so that they can be returned if nothing of significance 
     /// changed. The <see cref="VersionStamp"/> is produced by <see cref="Project.GetDependentVersionAsync(CancellationToken)"/> while the 
     /// <see cref="Checksum"/> is produced by <see cref="Project.GetDependentChecksumAsync(CancellationToken)"/>.  The former is faster
     /// and works well for us in the normal case.  The latter still allows us to reuse diagnostics when changes happen that
     /// update the version stamp but not the content (for example, forking LSP text).
     /// </summary>
-    private sealed class DiagnosticsPullCache(string uniqueKey) : VersionedPullCache<(int globalStateVersion, VersionStamp? dependentVersion), (int globalStateVersion, Checksum dependentChecksum), DiagnosticsRequestState, ImmutableArray<DiagnosticData>>(uniqueKey)
+    private sealed class DiagnosticsPullCache(string uniqueKey) : VersionedPullCache<CheapDiagnosticState, ExpensiveDiagnosticState, DiagnosticsRequestState, ImmutableArray<DiagnosticData>>(uniqueKey)
     {
-        public override async Task<(int globalStateVersion, VersionStamp? dependentVersion)> ComputeCheapVersionAsync(DiagnosticsRequestState state, CancellationToken cancellationToken)
+        public override async Task<CheapDiagnosticState> ComputeCheapVersionAsync(DiagnosticsRequestState state, CancellationToken cancellationToken)
         {
-            return (state.GlobalStateVersion, await state.Project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false));
+            var sgState = state.Project.Solution.GetSourceGeneratorExecutionVersion(state.Project.Id);
+            return new(state.GlobalStateVersion, sgState, await state.Project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false));
         }
 
-        public override async Task<(int globalStateVersion, Checksum dependentChecksum)> ComputeExpensiveVersionAsync(DiagnosticsRequestState state, CancellationToken cancellationToken)
+        public override async Task<ExpensiveDiagnosticState> ComputeExpensiveVersionAsync(DiagnosticsRequestState state, CancellationToken cancellationToken)
         {
-            return (state.GlobalStateVersion, await state.Project.GetDependentChecksumAsync(cancellationToken).ConfigureAwait(false));
+            var sgState = state.Project.Solution.GetSourceGeneratorExecutionVersion(state.Project.Id);
+            return new(state.GlobalStateVersion, sgState, await state.Project.GetDependentChecksumAsync(cancellationToken).ConfigureAwait(false));
```

Comment:
this was another place where we were storing the dependent checksum.  Which means we would incorrectly return stale diagnostics if only the source generated executio nversion changed.